# Classification Problem Lab
## Predict Stay/Leave The Company

Your boss was extremely happy with your work and decided to entrust you with a task. They've seen a lot of people leave the company recently and they would like to understand why that's happening. They have collected historical data on employees and they would like you to build a model that is able to predict which employee will leave next. The would like a model that is better than random guessing. They also prefer false negatives than false positives, in this first phase. Fields in the dataset include:

- Employee satisfaction level
- Last evaluation
- Number of projects
- Average monthly hours
- Time spent at the company
- Whether they have had a work accident
- Whether they have had a promotion in the last 5 years
- Department
- Salary
- Whether the employee has left (เป็นตัวแปรที่ต้องการทำนาย)

Your goal is to predict the binary outcome variable `left` using the rest of the data. Since the outcome is binary, this is a classification problem.

This dataset comes from https://www.kaggle.com/ludobenistant/hr-analytics/ and is released under [CC BY-SA 4.0 License](https://creativecommons.org/licenses/by-sa/4.0/).

#### Step 1 — Data Pre-processing (อ่าน data เข้ามา มีากรดูสถิติเกี่ยวกับ data เบื้องต้น มีการจัดการ column ที่เป็นข้อความให้เป็นตัวเลขหรือรูปแบบที่สามารถนำไปใช้ได้)

In [1]:
#import ....
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/TipGreenTea/MUICT-AST-IntroToAI_Datasets/main/HR_comma_sep.csv')
df

,satisfaction_level,last_evaluation,number_project,average_montly_hours,time_spend_company,Work_accident,promotion_last_5years,sales,salary,left
0,0.38,0.53,2,157,3,0,0,sales,low,1
1,0.80,0.86,5,262,6,0,0,sales,medium,1
2,0.11,0.88,7,272,4,0,0,sales,medium,1
3,0.72,0.87,5,223,5,0,0,sales,low,1
4,0.37,0.52,2,159,3,0,0,sales,low,1
...,...,...,...,...,...,...,...,...,...,...
14994,0.40,0.57,2,151,3,0,0,support,low,1
14995,0.37,0.48,2,160,3,0,0,support,low,1
14996,0.37,0.53,2,143,3,0,0,support,low,1
14997,0.11,0.96,6,280,4,0,0,support,low,1


In [ ]:
#CODE HERE
#สามารถตัด column sales, salary ออกจาก X ได้เลย
X = df.drop(['sales', 'salary', 'left'], axis=1)
Y = df['left']


#### Step 2 — Separating Your Training and Testing Datasets

In [ ]:
#CODE HERE
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)


In [14]:
X_train

array([[0.72527473, 0.703125  , 0.6       , ..., 0.375     , 0.        ,
        0.        ],
       [0.04395604, 0.859375  , 0.        , ..., 0.375     , 0.        ,
        0.        ],
       [0.63736264, 0.59375   , 0.2       , ..., 0.125     , 0.        ,
        0.        ],
       ...,
       [0.65934066, 0.90625   , 0.6       , ..., 0.        , 0.        ,
        0.        ],
       [0.75824176, 0.359375  , 0.6       , ..., 0.125     , 0.        ,
        0.        ],
       [0.51648352, 0.5       , 0.6       , ..., 0.        , 0.        ,
        0.        ]])

In [15]:
X_test

,satisfaction_level,last_evaluation,number_project,average_montly_hours,time_spend_company,Work_accident,promotion_last_5years
1670,0.44,0.57,2,141,3,0,0
13378,0.55,0.96,3,194,3,0,0
10233,0.72,0.67,5,210,2,0,0
4719,0.96,0.75,4,177,2,0,0
7003,0.96,0.54,3,198,3,0,0
...,...,...,...,...,...,...,...
14035,0.75,0.79,4,263,3,0,1
8469,0.92,0.49,5,178,2,1,0
12453,0.37,0.54,2,145,3,0,0
9431,0.52,0.61,4,163,2,0,0


#### Step 3 — Transforming the Data (Optional จะทำ Feature Scaling หรือไม่ก็ได้ ถ้าไม่ทำ สามารถข้าม step นี้ไปได้เลย)

In [ ]:
#CODE HERE
from sklearn.preprocessing import StandardScaler, MinMaxScaler
#sc = StandardScaler()  
sc = MinMaxScaler()
X_train = sc.fit_transform(X_train) #สร้างหน่วยที่จะแปลงแค่ dataset ชุดนี้
X_train

array([[0.72527473, 0.703125  , 0.6       , ..., 0.375     , 0.        ,
        0.        ],
       [0.04395604, 0.859375  , 0.        , ..., 0.375     , 0.        ,
        0.        ],
       [0.63736264, 0.59375   , 0.2       , ..., 0.125     , 0.        ,
        0.        ],
       ...,
       [0.65934066, 0.90625   , 0.6       , ..., 0.        , 0.        ,
        0.        ],
       [0.75824176, 0.359375  , 0.6       , ..., 0.125     , 0.        ,
        0.        ],
       [0.51648352, 0.5       , 0.6       , ..., 0.        , 0.        ,
        0.        ]])

#### Step 4 — Building ML models (สร้าง model และ train model จาก training data)

In [ ]:
#CODE HERE
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
model = LogisticRegression(random_state = 0)
#model = SVC()

In [6]:
model.fit(X_train, Y_train)

LogisticRegression(random_state=0)

#### Step 5 — Running Predictions on the Test Set (อธิบายว่าได้ accuracy, precision, recall เท่าไรให้กับ LA)

In [29]:
#CODE HERE
y_pred = model.predict(sc.transform(X_test))

In [30]:
from sklearn.metrics import classification_report
report = classification_report(Y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.80      0.92      0.86      2299
           1       0.49      0.25      0.33       701

    accuracy                           0.77      3000
   macro avg       0.65      0.59      0.59      3000
weighted avg       0.73      0.77      0.73      3000



**Accuracy**

In [31]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, y_pred)

0.765

Accuracy: 0.77 โดยภาพรวมแล้ว ถ้าให้โมเดลทายพนักงาน 100 คน โมเดลจะทายถูก (ไม่ว่าจะทายว่าออกหรือไม่ออก) จำนวน 77 คน "
ซึ่ง ฟังดูเหมือนดี แต่ต้องระวัง เพราะถ้าคนส่วนใหญ่ในบริษัท "ไม่ออก" โมเดลแค่ทายมั่วว่า "ทุกคนไม่ออก" คะแนนก็อาจจะสูงได้ (สังเกตจากบรรทัด support ที่ class 0 มีตั้ง 2299 คน แต่ class 1 มีแค่ 701 คน)

Precision (ความแม่นยำในการจับผิด) 0.49 "เมื่อไหร่ก็ตามที่โมเดลชี้หน้าพนักงานคนนึงแล้วบอกว่า 'คนนี้จะลาออกแน่ๆ!' ... โมเดลพูด ถูกแค่ 49% (เกือบครึ่งต่อครึ่ง) โมเดลเราขี้ระแวง ทายว่าออกเยอะไว้ก่อน แต่จริงๆ อาจจะไม่ออกก็ได้ (False Positive เยอะ)

Recall (ความครอบคลุมในการจับ) 0.25 "สมมติในความจริงมีพนักงานจะลาออก 100 คน... โมเดลเรา จับได้จริงแค่ 25 คน เท่านั้น (อีก 75 คนที่เหลือ โมเดลมองพลาดว่าเขาจะอยู่ต่อ)" โมเดลเรา "ตาถั่ว" พอสมควร ปล่อยคนจะลาออกหลุดรอดไปได้เยอะมาก (False Negative เยอะ)

Gemini
"โมเดลนี้มีความแม่นยำรวม (Accuracy) อยู่ที่ 77% แต่จุดอ่อนคือโมเดลยังทำนาย กลุ่มคนที่จะลาออก (Class 1) ได้ไม่ค่อยดีนัก โดยมีค่า Recall เพียง 25% แปลว่าเราจะพลาดโอกาสในการรั้งพนักงานที่จะลาออกไปถึง 3 ใน 4 คน สาเหตุหลักน่าจะมาจากข้อมูลที่ไม่สมดุล (Imbalanced Data) คือมีตัวอย่างคนลาออกให้โมเดลเรียนรู้น้อยเกินไปครับ (มีแค่ 701 คน เทียบกับคนอยู่ต่อ 2299 คน)"